kernel: Introduce a way to set minimum system workqueue stack size

jukkar opened the pull request:
Instead of user trying to figure out what is the system workqueue stack size in the system, let the various subsystems etc. specify their need using a Kconfig option. The build system will then check these values and use the largest one as a system workqueue stack size.

This works similar way as the heap size calculation introduced in commit 3fbf12487c6c01c488210bc56b0f342f07098df9 but instead of adding the values together it selects the largest one defined for the stack size.

jukkar committed:
kernel: Introduce a way to set minimum system workqueue stack size

Instead of user trying to figure out what is the system workqueue stack
size in the system, let the various subsystems etc. specify their need
using a Kconfig option. The build system will then check these values
and use the largest one as a system workqueue stack size.

This works similar way as the heap size calculation introduced
in commit 3fbf12487c6c01c488210bc56b0f342f07098df9 but instead of adding
the values together it selects the largest one defined for the stack
size.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: Add CONFIG_SYSTEM_WORKQUEUE_STACK_IGNORE_MIN to configs

As the tests are already set to run with a proper system workqueue
stack size, set the CONFIG_SYSTEM_WORKQUEUE_STACK_IGNORE_MIN=y to
their configs so that the correct value is being used.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
doc: migration-guide-4.4: Add entry about system workq stack size changes

The system workqueue stack size is now controlled by taking largest
value of the values of CONFIG_SYSTEM_WORKQUEUE_STACK_MIN_SIZE_* options.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
kernel: Create CONFIG_SYSTEM_WORKQUEUE_STACK_MIN_SIZE_APP option

Create a stack size option for application use. This can be used
as a default setting for those applications that do not want to
create their own Kconfig option.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt reviewed (dismissed):
Nice!

nordicjm commented on kernel/CMakeLists.txt:
```diff
@@ -167,6 +167,38 @@ if(${CONFIG_KERNEL_MEM_POOL})
+if(CONFIG_SYSTEM_WORKQUEUE_STACK_IGNORE_MIN)
+  set(final_systemwq_stack_size ${CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE})
+else()
+  # Import all custom SYSTEM_WORKQUEUE_STACK_ size requirements
```
*CONFIG_SYSTEM_WORKQUEUE_STACK_MIN_SIZE_

nordicjm commented on samples/Kconfig:
```diff
@@ -0,0 +1,8 @@
+
+config SYSTEM_WORKQUEUE_STACK_MIN_SIZE_SAMPLE
+	int "System workqueue stack size for a sample application"
+	default 0
```
don't really want to add a Kconfig file for samples - also note that this file is global, it will be present in all applications

jukkar commented on kernel/CMakeLists.txt:
```diff
@@ -167,6 +167,38 @@ if(${CONFIG_KERNEL_MEM_POOL})
+  endif()
+endif()
+
+zephyr_compile_definitions(SYSTEM_WORKQUEUE_STACK_SIZE=${final_systemwq_stack_size})
```
If I do the change, the build fails
```
zephyr/kernel/system_work_q.c:18:30: error: ‘SYSTEM_WORKQUEUE_STACK_SIZE’ undeclared here (not in a function); did you mean ‘CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE’?
   18 |                              SYSTEM_WORKQUEUE_STACK_SIZE);
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

nordicjm replied:
how about using zephyr_library_* and then adding a function to the .c file which returns the system workqueue size? I don't know what use people would have for this outside of the kernel file itself

tejlmand replied:
> You don't know that, there can be downstream code that does something with the CONFIG_ definition 🤷

but that still doesn't mean we should apply `SYSTEM_WORKQUEUE_STACK_SIZE` to all libs.
The fact that @jukkar did try to use `zephyr_library_compile_definitions()` does indicate the intention for this definition was only `system_work_q.c`, and not global appearance.

But as also mentioned, `kernel` is **not** a Zephyr library, so normal CMake functions must be used, see: https://github.com/zephyrproject-rtos/zephyr/pull/98058#discussion_r2451499626

> how about using zephyr_library_* and then adding a function to the .c file which returns the system workqueue size?

sound fine to me, but why not wait with that until we actually know that someone is going to need / use it ?
For now, I think we should consider `SYSTEM_WORKQUEUE_STACK_SIZE` as internal to the kernel lib.

jukkar committed:
modules: loramac-node: No need to depend on system workq size

The system work queue size is automatically set properly for
semtech driver when soft_se is enabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Rebased to main
* Fixed semtech lora driver dependency to system work queue size

aescolar commented on samples/Kconfig:
```diff
@@ -0,0 +1,8 @@
+
+config SYSTEM_WORKQUEUE_STACK_MIN_SIZE_SAMPLE
+	int "System workqueue stack size for a sample application"
+	default 1024
```
Note that as this file is always sourced, this will be a minimum in general for any app (test/sample or anything else).
(it seems off to me to create a new kconfig file for this at this level)
Maybe instead either:
* there could be some more generic SYSTEM_WORKQUEUE_STACK_MIN_SIZE_APP defined in kernel/Kconfig, 
* or CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE could default to this 1024 (To me it seems that defaulting CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE to 0 is wrong, as with 0 nothing can work)

?

jukkar commented:
Changes:
* Removed the bsim commit as it was not needed
* Removed the sample Kconfig file with `CONFIG_SYSTEM_WORKQUEUE_STACK_MIN_SIZE_SAMPLE` and replaced it with `CONFIG_SYSTEM_WORKQUEUE_STACK_MIN_SIZE_APP` as suggested by @aescolar

jukkar commented:
ping @nordicjm, please take a look

jhedberg commented on doc/releases/migration-guide-4.3.rst:
4.3 was already released. Please move this to the 4.4 migration guide.

nordicjm commented on tests/subsys/mgmt/mcumgr/smp_version/boards/qemu_riscv64.conf:
```diff
@@ -1 +1,2 @@
 CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096
+CONFIG_SYSTEM_WORKQUEUE_STACK_IGNORE_MIN=y
```
why? The calculated value is 1024, and the stack is being set to 4096, so I'm not understanding why ignoring a warning which is only printed if 4096 is less then 1024 (which isn't the case here) is needed?

jukkar commented on kernel/Kconfig:
```diff
@@ -621,6 +621,12 @@ config SYSTEM_WORKQUEUE_STACK_IGNORE_MIN
 	  is known for a given application.
 
+config SYSTEM_WORKQUEUE_STACK_MIN_SIZE_APP
+	int "System workqueue stack size for a applications"
```
For the samples, I assumed that we want to show how to use this new Kconfig option so did not use the `CONFIG_SYSTEM_WORKQUEUE_STACK_IGNORE_MIN` option in samples. So instead of creating a new Kconfig option for each sample that wants to set the stack size, I used one umbrella Kconfig option that can be used for this purpose. As only one application is run at a time, one Kconfig option should be enough for all the samples.

tejlmand reviewed (changes requested):
I like this idea, though I wonder why this is a min size, and not add size like with heap.

Though my main concern is taht if introducing a min size, then more subsystems should be updated to provide proper values depending on their configuration and not leave this for the application to find out.

If samples / applications anyway need to set a min size in many cases, then we could just keep existing `SYSTEM_WORKQUEUE_STACK_SIZE` approach.

JordanYates commented:
> I like this idea, though I wonder why this is a min size, and not add size like with heap.

Because only one work item can ever be running at one time. If a DNS work item needs a stack of 1024, SNTP needs 2048 and DHCP needs 1500, the system workqueue only needs to the 2048 bytes large to run all work items successfully.

This is different from the heap because all 3 subsystems could have memory allocated simultaneously from the heap, necessitating the "add" behaviour.

cfriedt commented:
I personally would love to see a common ways of summing, finding the minimum of, etc, distributed Kconfig options like this.

One thing I found kind of interesting, is this construct
```
CC := gcc

config CC_HAS_FOO
        def_bool $(shell, $(srctree)/scripts/gcc-check-foo.sh $(CC))
```
which evaluates to
```
config CC_HAS_FOO
        def_bool y
```

For reference, it's directly out of the kbuild docs
https://docs.kernel.org/kbuild/kconfig-macro-language.html#concept

So something like what is below isn't even that far-fetched
```
foo_size_min_pattern := CONFIG_FOO_SIZE_MIN_

config FOO_SIZE_MIN
        # evaluate the minimum of all CONFIG_FOO_SIZE_MIN_BAR, CONFIG_FOO_SIZE_MIN_BAZ, ...
        def_int $(shell, $(srctree)/scripts/kconfig_eval_min.py $(foo_size_min_pattern))
```

The tricky part is that it's a bit of a chicken / egg problem as @pdgendt has mentioned (e.g. could require running through Kconfig twice to take advantage of proper dependencies, etc).

cc @tejlmand

tejlmand commented:
> `foo_size_min_pattern := CONFIG_FOO_SIZE_MIN_`

@cfriedt problem is that this line means assign the string `CONFIG_FOO_SIZE_MIN_` to variable `foo_size_min_pattern`.

You want the value of `CONFIG_FOO_SIZE_MIN_`, but such value doesn't exists, and running Kconfig twice wouldn't help, it's still a string.

When we run Kconfig, then input (seen from Kconfig itself) is constant, meaning running Kconfig twice should not change the content of `.config`. If that is not honored in future, because we start to feed output of Kconfig into itself, then anything can start to depend on outcome of previous run. You cannot restrict such feature to only handle a specific use-case.
If we re-run Kconfig people will use and mis-use it for all sorts of tricks.

Which then can result in doing a third and fourth run  giving different results (incremental builds may trigger a Kconfig re-run).

This to me sounds like a recipe for disaster, compared to running Kconfig once, and then let CMake do post-calculation.

cfriedt commented:
@tejlmand - no.

In this case `CONFIG_FOO_SIZE_MIN_` is a simple literal passed to the shell script for the purpose of pattern matching. It's not magically doing anything.

What I'm implying is that (at most) one "subshell" invocation of `kconfiglib` in a script would be enough to generate a finalized `.config` that could be used to search for individual `CONFIG_FOO_SIZE_MIN_BAR=3` and `CONFIG_FOO_SIZR_MIN_BAZ=5` values (that match the pattern). Then, the script could sum them, and print the value `8` to stdout, thus ensuring that the `CONFIG_FOO_SIZE` value was assigned `8` in the primary `kconfiglib` invocation.

Given that `kconfig` typically evaluates in a fraction of a second (at least for Linux), the overhead should be pretty negligible.

An environment variable could be set and checked by the script to enforce the "maximally one" subshell limit.
